feat(controller): Add endpoint to regenerate tokens

Joshua-Anderson opened the pull request:
If a token was leaked, it could be regenerated with: `deis auth:regenerate`. Admin can regenerate a user's token with `deis auth:regenerate -u user` and could also regenerate everybody's auth token, in the event of a security breach, with `deis auth:regenerate --all=true`

carmstrong commented:
Not sure introducing a `deis token` command is the best way to expose this, as a. Maybe something like `deis auth:regenerate`? A user could do it to himself, or an admin could invoke the command on any user. 

I also wonder if `deis auth:logout` could suffice -- an admin could supply a username to the command, and it would log the user out by regenerating their access token.

bacongobbler commented:
`deis auth:token --new`, perhaps?

Joshua-Anderson commented:
I'm fine with either of those client commands. Is creating a new endpoint for the controller the right place to put this, or would you rather see it under, say, `/v1/auth/token/`?

mboersma commented:
I think `v1/auth/tokens` looks cleaner--we grouped the other auth URLs there (although they aren't proper REST endpoints). `v1/tokens` is acceptable too.

bacongobbler commented:
IOW what @mboersma is trying to say that the API implementation is really up to you, so use your own discretion. We're mostly concerned with getting the UX nailed down. Refactoring APIs is easy; refactoring UX is "impossible" unless we release v2.0, so that's what we're mostly concerned with :)

Joshua-Anderson commented:
I added actual code now. For now, I went with `auth:regenerate`. I also added `auth:regenerate --all=true` so a admin can regenerate everybody's token in the event of a security breach and `auth:regenerate -u user` so that a admin can regenerate a specific user.

bacongobbler commented on docs/managing_deis/operational_tasks.rst:
```diff
@@ -52,28 +52,22 @@ Re-issuing User Authentication Tokens
 
-At this point, let's re-issue an auth token for this user. Let's assume that the name for the user
-is Bob (poor Bob):
+A admin can also regenerate the token of another user like this:
```
An*

We should alsouse the fully-formed "administrator" here, since this is formal documentation. :necktie:

bacongobbler commented on docs/managing_deis/operational_tasks.rst:
```diff
@@ -83,14 +77,10 @@ token:
-re-issued a new token:
+They will need to log back in to use their new auth token.
+
+If there is a cluster wide security breach, a admin can regenerate everybody's auth token like this:
```
an administrator*

bacongobbler commented on docs/reference/api-v1.4.rst:
```diff
@@ -117,6 +118,40 @@ Example Response:
+
+.. code-block:: console
+
+    DELETE /v1/auth/tokens/ HTTP/1.1
```
Hmm... Maybe we should make this a POST instead of a DELETE. We're not actually deleting a token here, just generating a new one.

carmstrong replied:
:+1:

carmstrong commented:
> For now, I went with auth:regenerate. I also added auth:regenerate --all=true so a admin can regenerate everybody's token in the event of a security breach and auth:regenerate -u user so that a admin can regenerate a specific user.

:+1: Awesome work so far, @Joshua-Anderson!

bacongobbler commented:
note to self: investigate implications with external auth (we have support for LDAP)

Joshua-Anderson commented:
@bacongobbler I fixed the documentation and changed from DELETE to POST.

carmstrong commented:
@Joshua-Anderson Looks like we need a rebase now :(

Joshua-Anderson commented:
@carmstrong Done!

carmstrong commented:
Code LGTM. We should see what effect this has on the LDAP auth integration, if any.

mboersma commented:
Talking with @bacongobbler, we're confident LDAP isn't hampered by this.That's on the backend and ends up using the same token method where the client is concerned, so this should not matter.

Code LGTM.

bacongobbler commented on client/deis.py:
```diff
@@ -941,6 +942,36 @@ def auth_whoami(self, args):
+        Options:
+          -u --username=<username>
+            specify user to regenerate. Requires admin privilages.
+          --all=<all>
```
We should change this to `--all` so it'll end up being true if set, otherwise users will have to use `--all=foo` in order to use this feature flag.

bacongobbler commented on controller/api/permissions.py:
```diff
@@ -129,3 +129,18 @@ def has_permission(self, request, view):
+        if 'username' in request.data or 'all' in request.data:
+            return request.user.is_superuser
+        else:
+            return True
```
we should always be checking if the user is a superuser. In that case we can re-use the IsAdmin class instead.

bacongobbler replied:
actually I see the reason behind this now. Disregard!

bacongobbler commented on controller/api/views.py:
```diff
@@ -51,6 +52,40 @@ def passwd(self, request, **kwargs):
+        obj = self.get_object()
+
+        if 'all' in request.data:
+            for user in User.objects.exclude(username='AnonymousUser'):
```
rather than checking for the username, we can check `.is_anonymous()` instead to achieve the same effect. That may be more portable.

bacongobbler commented on controller/api/views.py:
```diff
@@ -51,6 +52,40 @@ def passwd(self, request, **kwargs):
+            for user in User.objects.exclude(username='AnonymousUser'):
+                token = Token.objects.get(user=user)
+                token.delete()
+                token = Token.objects.create(user=user)
```
let's remove the assignment since we're not doing anything with it.

bacongobbler commented on controller/api/views.py:
```diff
@@ -51,6 +52,40 @@ def passwd(self, request, **kwargs):
+                token.delete()
+                token = Token.objects.create(user=user)
+            return Response("",
+                            status=status.HTTP_204_NO_CONTENT,
```
this should be a 200 OK response to confirm that the operation succeeded

bacongobbler commented on controller/api/tests/test_auth.py:
```diff
@@ -278,3 +278,33 @@ def test_change_user_passwd(self):
+        response = self.client.post(url, '{"all" : "true"}',
+                                    content_type='application/json',
+                                    HTTP_AUTHORIZATION='token {}'.format(self.admin_token))
+        self.assertEqual(response.status_code, 204)
```
same as other comment, we should be testing for a 200 OK response

Joshua-Anderson commented:
@bacongobbler Thanks for the feedback! I think I addressed it.

Joshua-Anderson commented:
The tests seemed to time out, could somebody please rerun this PR.

mboersma commented:
@Joshua-Anderson could you rebase this off current master? I think #3771 was killing most of the tests until we fixed it. Apologies for the constant rebasing. :hammer:

Joshua-Anderson commented:
@mboersma Rebased!

carmstrong commented:
Code LGTM.

bacongobbler commented:
Much better! Sadly we're in code freeze atm for v1.7 so I'll defer the merge to @carmstrong but code LGTM :)

Joshua-Anderson commented:
I bumped the api version since v1.4 was released.

mboersma commented:
Code LGTM after updating the API to v1.5.

carmstrong commented:
Code LGTM.
